util/gogc.go

```diff
@@ -40,10 +40,64 @@ func SetGOGC(val int) {
 	}
 	debug.SetGCPercent(val)
 	metrics.GOGC.Set(float64(val))
-	atomic.StoreInt64(&gogcValue, int64(val))
+	gogcValue.Store(int64(val))
 }
 
 // GetGOGC returns the current value of GOGC.
 func GetGOGC() int {
-	return int(atomic.LoadInt64(&gogcValue))
+	return int(gogcValue.Load())
+}
+
+// GOGCStatSampler is a sampler for GOGC.
+type GOGCStatSampler struct {
+	last struct {
+		now          int64
+		lastGC       time.Time
+		gcPauseTotal uint64
+	}
+	GCRunCount uint64
+	exitChan   chan struct{}
+}
+
+// NewRuntimeStatSampler returns a sampler for GOGC.
+func NewRuntimeStatSampler() *GOGCStatSampler {
+	return &GOGCStatSampler{
+		exitChan: make(chan struct{}),
+	}
+}
+
+var gc debug.GCStats
+
+func (gss *GOGCStatSampler) sample() {
+	debug.ReadGCStats(&gc)
+	now := time.Now().UnixNano()
+	dur := float64(now - gss.last.now)
+	gcPauseRatio := float64(uint64(gc.PauseTotal)-gss.last.gcPauseTotal) / dur
```

Comment:
If you want to know `GCPausePercent` as a guage metric, you don't need to record the `last.gcPauseTotal`
Just use: time.Since(program start) / gc.PauseTotal

Reply:
it is copied from CRDB https://github.com/cockroachdb/cockroach/blob/e5eda9f716a6e1a30f17adaf96de953187030ff8/pkg/server/status/runtime.go#L523. I think this is to show the rate of change of GC duration.